ReadMe.txt
GQR, a Fast Reasoner for Binary Qualitative Constraint Calculi
--------------------------------------------------------------

GQR accepts definitions of qualitative calculi like RCC-8 or Allen's
interval algebra and determines whether constraint networks in those calculi
are algebraically closed or globally consistent. It furthermore supports
building combinations of calculi based on existing calculi definitions.

The source code is documented via doxygen. By typing in "gqr help",
the program gives a brief description of its call parameters.

Examples of calculi and constraint network files can be found in data/.

If you have any questions regarding GQR, feel free to ask them by e-mail to
Matthias Westphal <westpham@informatik.uni-freiburg.de>,
Zeno Gantner <gantner@ismll.uni-hildesheim.de> or
Stefan Wölfl <woelfl@informatik.uni-freiburg.de>.


Some information about GQR can be found in

Zeno Gantner, Matthias Westphal, Stefan Wölfl (2008): GQR - A Fast
Reasoner for Binary Qualitative Constraint Calculi, to appear in
Workshop Notes of the AAAI-08 Workshop on Spatial and Temporal
Reasoning, Chicago, USA, AAAI.

If you use GQR in your own research, we encourage you to cite this article.


Installation
------------

For installation, you need a C++ compiler (tested with g++ 4.3.3) and Python.

To build and run GQR after download, extract the tar file, change to the new
directory and type:

a) ./waf configure

Configures GQR and checks the necessary dependencies.

b) ./waf

Compiles GQR.

c) ./_build_/default/gqr/gqr

Runs the GQR binary. During runtime GQR requires access to the calculi
definition files found in "data/". GQR searches for the calculi definition
files in the directory specified by the environment variable "GQR_DATA_DIR", in
the directory "data/" in the current working directory, and finally in the directory
given during configure by --prefix and --data-dir (in this specific order).

Additionally you can type "./waf check" to perform some unit tests.

GQR uses the waf build system. Further options for compilation and system-wide
installation are as follows:

./waf --help:  for usage page

./waf configure  : configuration step of GQR build
                   --relation-size=N sets the maximum number of base relations
                   to N
                   --enable-xml enables support for calculi definitions in XML
./waf            : build GQR
./waf install    : install GQR into your system;
                   check variables --prefix and
                   --data-dir in the configuration step;
./waf check      : build and run unit tests
./waf uninstall  : 
./waf clean      :
./waf distclean  :
./waf dist       : create a tarball of the current source tree


C Library
---------

The main GQR functions can be compiled into a C library which can be accessed from other
programming languages.

./waf configure --enable-libgqr

enables the library during configure. "./waf" then builds the library binary in
"_build_/default/gqr/". C example code that uses the library can be found in
"gqr/libgqr_example/". "gqr/libgqr_example/ReadMe.txt" contains further information.

Additionally, a python class using ctypes can be found in "gqr/python/".


Acknowledgements
----------------

This program was initially developed by Zeno Gantner under the supervision
of Malte Helmert, Bernhard Nebel and Stefan Wölfl at University of Freiburg.

Zeno would like to thank Jochen Eisinger and Matthias Spiller for their
occasional help in his struggle with C++, and Malte Helmert for introducing
him into the arcane arts of writing Makefiles and using the gprof profiler.

We would also like to thank Bernhard Nebel and Peter van Beek for giving
permission to use their queue implementation and to distribute it under the
terms of the GNU General Public License.

Matthias Westphal wrote and integrated the current RelationFixedBitSet
implementation, as well as all search improvements.

Further work on data structures and major parts of the code refactoring was
done by Isgandar Valizada and Thomas Liebetraut.

License
-------

Copyright (C) 2006, 2007, 2008, 2010 Zeno Gantner, Stefan Wölfl
Copyright (C) 2006, 2007 Micha Altmeyer
Copyright (C) 2007-2012 Matthias Westphal

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; see the file COPYING.  If not, write to the Free
Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
USA.

GQR contains other sections of code that are copyrighted by other
institutions or individuals. The GQR copyright does not override or
invalidate those copyrights.



Changelog
Changes since release 1418:
- Major code refactoring
- Search with nogoods and restarts
- Speed and scalability improvements

Changes since release 1417:
- Ability to compile GQR into a C library (based on the GObject framework)
- Minor bugfixes for gcc-4.5, data search path, memory leaks, unit tests and waf

Changes since release 1298:
- Automatically compute the splitting of relations into tractable relations on startup, instead of relying on precomputed files
- Enforce total order on relations within WeightValueOrder
- Replace std::priority_queue with heap-based queue that supports decrease-key operations in weighted algebraic closure
- Remove std::tr1 dependency; GQR now ships with a HashSet class
- Minor bugfixes for input file handling

Changes since release 1200:
- Multiple relation data types precompiled in binary
- Use trailing-like history to reduce memory consumption within search
- Enable weighted a-closure by default
- Bugfix for heuristic in weighted a-closure
- Remove multiple heuristic edge queue type
- Make non-precomputed composition use the size-trick; as found in the cache precomputation
- Update the SWIG functionality

Changes since release 1137:
- reduce memory consumption within the path consistency algorithm
- new precomputation architecture for calculi operations (improves constraint propagation speed)

Changes since release 1089:
- GCC 4.4.1 compile fix
- minor search space reduction in 2-way branching DFS with dynamic edge selection

Changes since release 729:
- switched to subcommand commandline structure (subversion style)
- support for combinations of binary constraint calculi
- correctly handle calculi whose base relations are non-serial
- speedup for weighted path consistency scheme
- more edge ordering heuristics
- dynamic edge ordering heuristics
- new default heuristic (dynamic) "weight over learned weights" (a variant of "domain over weighted degree")
- value ordering according to weight of relations
- support for eligible edges (Condotta et al.)
- support for MINCSP calculation

Changes since version 20102006:
- now uses waf (http://code.google.com/p/waf/) as build system
- added support for x86-64 CPUs (should also work on other 64 bit systems)
- new class RelationFixedBitset to represent relations as binary vectors
- added the cardinal directions calculus (cd)
- introduced some variable ordering heuristics to the backtracking search
- removed GUI code and Python bindings (will come back in the future)
- fixed composition table of the occlusion calculus
- fixed the cover files of the RCC-8 subclasses
- lots of bugfixes


gqr/gqr.cpp
// -*- C++ -*-

// Copyright (C) 2006, 2007, 2008 Zeno Gantner, Stefan Woelfl
// Copyright (C) 2006, 2007 Micha Altmeyer
// Copyright (C) 2007-2012 Matthias Westphal
// Copyright (C) 2011 Thomas Liebetraut
//
// This file is part of the Generic Qualitative Reasoner GQR.  This
// program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; see the file COPYING.  If not, write to
// the Free Software Foundation, 51 Franklin Street, Fifth Floor,
// Boston, MA 02110-1301, USA.

#include <cassert>
#include <iostream>
#include <string>
#include <limits.h>

#include <cstdlib>



//#include "commands/SubcommandConsistency.h"
//#include "commands/SubcommandMinCsp.h"
//#include "commands/SubcommandMultiConsistency.h"
//#include "commands/SubcommandMultiPathConsistency.h"

#include "gqr.h"

#include "commands/SubcommandAbstract.h"
#include "commands/SubcommandHelp.h"
#include "commands/SubcommandVersion.h"
#include "commands/SubcommandCheck.h"
#include "commands/SubcommandPathConsistency.h"
#include "commands/SubcommandConsistency.h"

/**
 * Generic solver for binary constraint networks.
 *
 * @see USAGE_STRING
 */

static SubcommandAbstract* getSubcommand(const int argc, char *argv[]) {
	// if there was no command specified, we want the help command to be
	// executed

	std::vector<std::string> args;
	for (int i = 0; i < argc; i++)
		args.push_back(std::string(argv[i]));

	std::string s;
	if (args.size() >= 2) // options given
		s = args[1];
	// The first argument determines the type of the subcommand
	// Then delegate the parameter parsing to the specific command


	if (s == "help")
		return new SubcommandHelp(args);
	if (s == "version")
		return new SubcommandVersion(args);
	if (s == "check")
		return new SubcommandCheck(args);
	if (s == "pc" || s == "path consistency")
		return new SubcommandPathConsistency(args);
	if (s == "c" || s == "consistency")
		return new SubcommandConsistency(args);

/*
//	Disabled for now

	if (s == "mincsp" || s == "min csp")
		return new SubcommandMinCsp(args);

	if (s == "mpc" || s == "multi path consistency")
		return new SubcommandMultiPathConsistency(args);

	if (s == "mc" || s == "multi consistency")
		return new SubcommandMultiConsistency(args);
*/

	return new SubcommandHelp(args);
}

int main(int argc, char* argv[])
{
	SubcommandAbstract *cmd = getSubcommand(argc, argv);
	const int ret = cmd->run();
	delete cmd;

	return ret;
}
